### PR TITLE
Add variant of new load and save ops for storing model params in a single file

### PR DESCRIPTION
**Pull request by sidgoyal78**
This PR basically implements a different approach for issue #7722 , and is based on the suggestions in PR #7780.

Things added:
- load_combine_op
- save_combine_op
- A unit test for saving and loading multiple LoDTensors
- A unit test for saving and loading 1 LoDTensor (so as to verify that it performs the function of the original load_op and save_op as well).

TODO: To remove the original load_op and save_op and rename these.

### CODE REVIEW

File: paddle/operators/load_combine_op.cc [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+  void Run(const framework::Scope &scope,
+           const platform::Place &place) const override {
+    auto filename = Attr<std::string>("file_path");
+
+    std::ifstream fin(filename);
+    PADDLE_ENFORCE(static_cast<bool>(fin),
+                   "Cannot open file %s for load_combine op", filename);
+
+    auto out_var_names = Outputs("Out");
+    PADDLE_ENFORCE(out_var_names.size(), "Output variables cannot be found");
```
Comment (Xreki) [inline paddle/operators/load_combine_op.cc]: Change to `PADDLE_ENFORCE_LT(out_var_names.size(), 0, "The number of output variables should be > 0.");`

File: paddle/operators/load_combine_op.cc [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    for (size_t i = 0; i < out_var_names.size(); i++) {
+      auto *out_var = scope.FindVar(out_var_names[i]);
+
+      PADDLE_ENFORCE(out_var != nullptr, "Output variable %s cannot be found",
+                     out_var_names[i]);
+
+      auto *tensor = out_var->GetMutable<framework::LoDTensor>();
+
+      uint64_t data_length;
+      char *buffer = NULL;
```
Comment (Xreki) [inline paddle/operators/load_combine_op.cc]: I think there is no need a buffer, we can directly use `fin` to read the file sequentially and deserialize the LoDTensor one by one.

File: paddle/operators/load_combine_op.cc [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+      buffer = new char[data_length];
+
+      // Read the serialized data into the buffer
+      fin.read(buffer, data_length);
+
+      std::string current_serialized_data;
+      current_serialized_data.assign(buffer, data_length);
+
+      // Create an input string stream
+      std::istringstream ist(current_serialized_data);
```
Comment (Xreki) [inline paddle/operators/load_combine_op.cc]: No need another stream, we can directly use `fin`.

File: paddle/operators/load_combine_op.cc [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+    for (size_t i = 0; i < out_var_names.size(); i++) {
+      auto *out_var = scope.FindVar(out_var_names[i]);
+
+      PADDLE_ENFORCE(out_var != nullptr, "Output variable %s cannot be found",
+                     out_var_names[i]);
+
+      auto *tensor = out_var->GetMutable<framework::LoDTensor>();
+
+      uint64_t data_length;
```
Comment (Xreki) [inline paddle/operators/load_combine_op.cc]: I am not sure is there still a need to save the `data_length`...

File: paddle/operators/load_combine_op.cc (hunk 1 above)
Comment (sidgoyal78) [reply paddle/operators/load_combine_op.cc]: Done. (I think you mean `PADDLE_ENFORCE_GT` :) )

File: paddle/operators/load_combine_op.cc (hunk 4 above)
Comment (sidgoyal78) [reply paddle/operators/load_combine_op.cc]: Yes, thanks for pointing this out. I have simplified the logic.

File: paddle/operators/load_combine_op.cc (hunk 2 above)
Comment (sidgoyal78) [reply paddle/operators/load_combine_op.cc]: Done, thanks for the suggestion.

File: paddle/operators/load_combine_op.cc (hunk 3 above)
Comment (sidgoyal78) [reply paddle/operators/load_combine_op.cc]: Done.

File: paddle/operators/load_combine_op.cc [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        .AsDuplicable();
+    AddAttr<std::string>("file_path",
+                         "(string) "
+                         "Variable will be load_combined from \"file_path\".")
+        .AddCustomChecker(
+            [](const std::string &path) { return !path.empty(); });
+    AddComment(R"DOC(
+LoadCombine Operator.
+
+LoadCombine operator combines together various tensor variable into a file.
```
Comment (kexinzhao) [inline paddle/operators/load_combine_op.cc]: Should the comment be something like "load together various tensor variable from a file"?

File: paddle/operators/save_load_op_new_test.cc [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#include "gtest/gtest.h"
+#include "paddle/framework/op_registry.h"
+
+USE_NO_KERNEL_OP(save);
```
Comment (kexinzhao) [inline paddle/operators/save_load_op_new_test.cc]: save -> save_combine?

File: paddle/operators/save_load_op_new_test.cc [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+USE_NO_KERNEL_OP(load);
```
Comment (kexinzhao) [inline paddle/operators/save_load_op_new_test.cc]: load -> load_combine?

File: paddle/operators/load_combine_op.cc (hunk 5 above)
Comment (kexinzhao) [inline paddle/operators/load_combine_op.cc]: Should it be something like "load various tensor variables from a file"?

Comment (sidgoyal78) [reply paddle/operators/load_combine_op.cc]: Yes, will change the comment.

File: paddle/operators/save_load_op_new_test.cc (hunk 6 above)
Comment (sidgoyal78) [reply paddle/operators/save_load_op_new_test.cc]: Fixed, thanks.

File: paddle/operators/save_load_op_new_test.cc (hunk 7 above)
Comment (sidgoyal78) [reply paddle/operators/save_load_op_new_test.cc]: Fixed, thanks. Based on the naming issues, i fixed some other things as well. Thanks.

File: paddle/operators/load_combine_op.cc (hunk 5 above)
Comment (sidgoyal78) [reply paddle/operators/load_combine_op.cc]: Done

File: paddle/operators/load_combine_op.cc [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+      }
+    }
+  }
+};
+
+class LoadCombineOpProtoMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  LoadCombineOpProtoMaker(OpProto *proto, OpAttrChecker *op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddOutput("Out", "(LoDTensor) The tensor need to be load_combined")
```
Comment (Xreki) [inline paddle/operators/load_combine_op.cc]: (LoDTensor) The tensor need to be load_combined -> (vector<LoDTensor>) The tensors need to be loaded

File: paddle/operators/load_combine_op.cc [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+        .AsDuplicable();
+    AddAttr<std::string>("file_path",
+                         "(string) "
+                         "Variable will be load_combined from \"file_path\".")
```
Comment (Xreki) [inline paddle/operators/load_combine_op.cc]: Variable will be load_combined from \"file_path\". -> Variables will be loaded from \"file_path\".
or The file_path to load Variables.

File: paddle/operators/load_combine_op.cc [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+        .AddCustomChecker(
+            [](const std::string &path) { return !path.empty(); });
+    AddComment(R"DOC(
+LoadCombine Operator.
+
+LoadCombine operator loads tensor variables from a file.
```
Comment (Xreki) [inline paddle/operators/load_combine_op.cc]: More comments here.

File: paddle/operators/save_combine_op.cc [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+      PADDLE_THROW("%s exists!, cannot save_combine to it when overwrite=false",
+                   filename, overwrite);
+    }
+
+    MkDirRecursively(DirName(filename).c_str());
+    std::ofstream fout(filename);
+    PADDLE_ENFORCE(static_cast<bool>(fout), "Cannot open %s to write",
+                   filename);
+
+    auto inames = Inputs("X");
```
Comment (Xreki) [inline paddle/operators/save_combine_op.cc]: inames -> in_names

File: paddle/operators/save_combine_op.cc [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+    PADDLE_ENFORCE_GT(
+        static_cast<int>(inames.size()), 0,
+        "The number of output variables should be greater than 0");
```
Comment (Xreki) [inline paddle/operators/save_combine_op.cc]: output -> input

File: paddle/operators/save_load_op_new_test.cc [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#include "gtest/gtest.h"
+#include "paddle/framework/op_registry.h"
+
+USE_NO_KERNEL_OP(save_combine);
+USE_NO_KERNEL_OP(load_combine);
+
+TEST(SaveLoadNewOp, CPU) {
```
Comment (Xreki) [inline paddle/operators/save_load_op_new_test.cc]: Move this unittest to save_load_combine_op_test.cc

File: paddle/operators/save_load_combine_op_test.cc [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+  }
+  return expect;
+}
+
+paddle::framework::LoDTensor* generate_placeholder_before_load(
+    const std::string out_var_name, paddle::framework::Scope& scope) {
+  auto load_var = scope.Var(out_var_name);
+  auto target = load_var->GetMutable<paddle::framework::LoDTensor>();
+  return target;
+}
```
Comment (Xreki) [inline paddle/operators/save_load_combine_op_test.cc]: Add a blank line here.

File: paddle/operators/save_load_combine_op_test.cc [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+#include <iostream>
+#include <string>
+#include <vector>
+#include "gtest/gtest.h"
+#include "paddle/framework/op_registry.h"
+
+USE_NO_KERNEL_OP(save_combine);
+USE_NO_KERNEL_OP(load_combine);
+
+int* create_for_save_combine_op(int x, int y, const std::vector<int>& lod_info,
```
Comment (Xreki) [inline paddle/operators/save_load_combine_op_test.cc]: create_for_save_combine_op -> CreateSaveCombineOp
the same for following functions.

File: paddle/operators/load_combine_op.cc (hunk 8 above)
Comment (sidgoyal78) [reply paddle/operators/load_combine_op.cc]: Done, thanks.

File: paddle/operators/load_combine_op.cc (hunk 9 above)
Comment (sidgoyal78) [reply paddle/operators/load_combine_op.cc]: Done

File: paddle/operators/load_combine_op.cc (hunk 10 above)
Comment (sidgoyal78) [reply paddle/operators/load_combine_op.cc]: Added.

File: paddle/operators/save_combine_op.cc (hunk 11 above)
Comment (sidgoyal78) [reply paddle/operators/save_combine_op.cc]: Modified, to make it consistent with variable name in load_combine: I modified it to `inp_var_names`

File: paddle/operators/save_combine_op.cc (hunk 12 above)
Comment (sidgoyal78) [reply paddle/operators/save_combine_op.cc]: Done.

File: paddle/operators/save_load_combine_op_test.cc (hunk 15 above)
Comment (sidgoyal78) [reply paddle/operators/save_load_combine_op_test.cc]: Done

File: paddle/operators/save_load_combine_op_test.cc (hunk 14 above)
Comment (sidgoyal78) [reply paddle/operators/save_load_combine_op_test.cc]: Done

File: paddle/operators/save_load_op_new_test.cc (hunk 13 above)
Comment (sidgoyal78) [reply paddle/operators/save_load_op_new_test.cc]: Done

File: paddle/operators/save_load_combine_op_test.cc [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+
+int* GetValuesAfterLoadCombineOp(paddle::framework::LoDTensor* target,
+                                 paddle::framework::Scope& scope,
+                                 paddle::framework::LoD& actual_lod) {
+  int* actual = target->data<int>();
+  actual_lod = target->lod();
+  return actual;
+}
+
+void check_values(int* expect, int* actual, paddle::framework::LoD expect_lod,
```
Comment (kexinzhao) [inline paddle/operators/save_load_combine_op_test.cc]: check_values -> CheckValues

Comment (sidgoyal78) [reply paddle/operators/save_load_combine_op_test.cc]: Done

**Pull request merged**